### PR TITLE
Hybrid updates

### PR DESCRIPTION
**Pull request by JackTemaki**
- support for stm and glm in via data (used for Switchboard setup)
 - NnSystem manages native op compilation
 - small refactoring of evaluation preparation
 - removes compute perplexity
 
 Note: Any recognition args now need to have `"native_ops": ["NativeLstm2"]` to get the old behavior.

### CODE REVIEW

File: common/setups/rasr/nn_system.py [hunk 1]
```diff
@@ -66,3 +66,34 @@ def __init__(
...
+
+        self.native_ops = {}  # type: Dict[str, tk.Path]
+
+    def compile_native_op(self, op_name: str):
+        """
+        Compiles and stores a native op for RETURNN
+
+        :param op_name: op name, e.g. "NativeLstm2"
+        """
+        native_lstm_job = returnn.CompileNativeOpJob(
```
Comment (vieting) [inline common/setups/rasr/nn_system.py:78]:
```suggestion
        native_op_job = returnn.CompileNativeOpJob(
```
Same applies a few lines down.

File: common/setups/rasr/rasr_system.py [hunk 2]
```diff
@@ -165,32 +166,45 @@ def _set_scorer_for_corpus(self, eval_corpus_key: str):
...
                 else dict(mapping={"[SILENCE]": ""})
             )
             self.set_kaldi_scorer(
                 corpus=eval_corpus_key,
                 **scorer_args,
             )
         elif self.rasr_init_args.scorer == "hub5":
-            self.set_hub5_scorer(corpus=eval_corpus_key)
+            scorer_args = (
+                scorer_args if self.rasr_init_args.scorer_args is not None else {}
```
Comment (vieting) [inline common/setups/rasr/rasr_system.py:182]:
```suggestion
                scorer_args if scorer_args is not None else {}
```
Would be more consistent with the change above.

File: common/setups/rasr/nn_system.py (hunk 1 above)
Comment (vieting) [inline common/setups/rasr/nn_system.py:78]:
```suggestion
        native_op_job = returnn.CompileNativeOpJob(
```
Same applies a few lines down.

File: common/setups/rasr/gmm_system.py [hunk 3]
```diff
@@ -1469,22 +1469,7 @@ def run(self, steps: Union[List[str], RasrSteps]):
...
-            self.create_stm_from_corpus(eval_c, **stm_args)
-            self._set_scorer_for_corpus(eval_c)
-
-            ppl_job = lm.ComputePerplexityJob(
-                self.crp[eval_c],
-                corpus_recipes.CorpusToTxtJob(
-                    self.crp[eval_c].corpus_config.file
-                ).out_txt,
-            )
-            tk.register_output(f"lms/{eval_c}.ppl", ppl_job.perplexity)
```
Comment (vieting) [inline common/setups/rasr/gmm_system.py:1487]: Is it intended to remove this? It does not seem to be moved somewhere else.

**Review by vieting — CHANGES_REQUESTED** [review]
Just some small comments.

File: common/setups/rasr/hybrid_system.py [hunk 4]
```diff
@@ -90,11 +90,13 @@ def __init__(
         self.cv_corpora = []
         self.devtrain_corpora = []
 
-        self.train_input_data = None
-        self.cv_input_data = None
-        self.devtrain_input_data = None
-        self.dev_input_data = None
-        self.test_input_data = None
+        self.train_input_data = None  # type:Optional[Dict[str, ReturnnRasrDataInput]]
```
[the remark this answers is not in the document]
Comment (JackTemaki) [reply common/setups/rasr/hybrid_system.py:93]: Yes, and I thought about adding that, but I did not want to mix it in here yet. So far we anyway have no other input that works properly. We can extend this here once we have that.

**Review by christophmluscher — APPROVED** [review]

**Review by vieting — APPROVED** [review]

**Pull request merged**